core/src/main/scala/kafka/log/Log.scala

```diff
@@ -893,6 +926,7 @@ class Log(@volatile var dir: File,
         activeSegment.truncateTo(targetOffset)
         updateLogEndOffset(targetOffset)
         this.recoveryPoint = math.min(targetOffset, this.recoveryPoint)
+        this.logStartOffset = math.min(targetOffset, this.logStartOffset)
```

Comment:
If everything is implemented correctly then log start offset should not decrease. But in reality we have seen unlean leader election during double failure where the logEndOffset is truncated beyond high watermark. It seems safer to preserve the existing behavior of `TruncateTo()` and allow logStartOffset to be reduced if such truncation does happen instead of throwing exception. Note that such truncation will only happen as part of unclean leader election initiated by the brokers themselves.